### ISSUE TITLE
Non-usascii test failing in Arch Linux, CentOS, FreeBSD

### ISSUE THREAD

**Issue by essen** [ninenines/erlang.mk#884]
Probably something to do with environment but it's difficult to pinpoint what exactly... Where it fails it fails with:

```
make[3]: *** No rule to make target '/root/erlang.mk/test/test_core_makedep_non_usascii_paths/hÃ©hÃ©/my_app/deps/my_dep/include/hello.hrl', needed by 'src/hello.erl'.  Stop.
make[2]: *** [erlang.mk:5176: app] Error 2
make[2]: Leaving directory '/root/erlang.mk/test/test_core_makedep_non_usascii_paths/héhé/my_app'
```

Note that the target has a bad name while Make enters the directory with a good name?

It fails in these environments, which are fairly bare bone. My laptop with Arch Linux doesn't fail:

https://buildkite.com/ninenines/erlang-dot-mk/builds/262

/cc @dumbbell

**Comment by crownedgrouse** [comment]
your test/core_makedep.mk is UTF-8 encoded, but result of mkdir seems to be Latin1. 
A check of $LANG  or `locale charmap` needed to set NON_USASCII_DIR variable in right encoding via a `iconv -f UTF-8 -t $(locale charmap)`  may help.
Note that make use `/bin/sh` by default underneath. It is maybe not same encoding than parent's shell. Considere set MAKESHELL current $SHELL to be sure having same environment.

**Comment by essen** [comment]
`$LANG` or `locale charmap` doesn't help. Maybe it's the shell I'll check later.

**Comment by essen** [comment]
```
arch linux:	bash in sh mode (5.0)				ERROR
alpine:		ash								OK
centos:		bash in sh mode (4.2)				ERROR
debian:		dash								OK
freebsd:		freebsd's sh						ERROR
macos:		bash in sh mode (3.2)				OK
ubuntu:		dash								OK
win10:		msys2-bash (in sh mode?) (4.4)		OK
```

So we can of course ignore Windows 10 since it's just a compatibility layer and obviously doesn't work the same. MacOS is an old version and might be custom because locales don't work quite the same, I don't know. I can't ssh and type `héhé` for some reason while it works in other platforms. I also get this for example:

```
minios:erlang.mk essen$ ls -l test/test_core_makedep_non_usascii_paths/
total 0
drwxr-xr-x  4 essen  staff  128 Apr 13 14:44 h??h??
```

Make version doesn't seem to matter.

/cc @dumbbell

**Comment by crownedgrouse** [comment]
Did you try to run
`LC_ALL=fr_FR.UTF-8 make ... `

**Comment by essen** [comment]
Arch Linux only has `C` and `POSIX` for one. Locales vary wildly depending on environments. This shouldn't make the tests fail.

**Comment by crownedgrouse** [comment]
What gives you
```
cd test/test_core_makedep_non_usascii_paths
ls * > test
file test
```
Also fix you try to surround mkdir by a $(shell ...) call instead mkdir directly?

**Comment by essen** [comment]
Says ASCII text in both working and non-working environments.

It's `my_app.d` file that's being generated wrong. Should be easy to figure out now.

**Comment by essen** [comment]
Good one has `é` encoded as `233`, bad one has `é` encoded as `195,169`.

**Comment by essen** [comment]
I think it's because of the old shell.

```
$ erl
Erlang/OTP 22 [erts-10.6] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe]

Eshell V10.6  (abort with ^G)
1> "hé\0".
[104,233,0]
^C^C
$ erl -oldshell
Eshell V10.6  (abort with ^G)
1> "hé\0".
[104,195,169,0]
```

**Comment by essen** [comment]
Thinking of something similar to this:

``` patch
-       ok = file:write_file("$(1)", unicode:characters_to_binary([
+       Output0 = [
                "# Generated by Erlang.mk. Edit at your own risk!\n\n",
                [[F, "::", [[" ", D] || D <- Deps], "; @touch \$$@\n"] || {F, Deps} <- Depend],
                "\nCOMPILE_FIRST +=", [[" ", TargetPath(CF)] || CF <- CompileFirst], "\n"
-       ])),
+       ],
+       Output = case "é" of
+               [233] -> unicode:characters_to_binary(Output0);
+               _ -> Output0
+       end,
+       ok = file:write_file("$(1)", Output),
```

**Comment by essen** [comment]
Fixed in 44b3c70a9ee51410aa224dae183fcf9436f01139. Thanks for the help.
